src/session.ts

```diff
@@ -0,0 +1,294 @@
+      }
+      this._kernel.start(success.data.kernel);
+      this._handleStatus('kernelCreated');
+    }, (error: IAjaxError) => {
+      this._handleStatus('kernelDead');
+    });
+  }
+
+  /**
+   * GET /api/sessions/[:session_id]
+   *
+   * Get information about a session.
+   *
+   */
+  getInfo(): Promise<ISessionId> {
+    return utils.ajaxRequest(this._sessionUrl, {
+      method: "GET",
+      dataType: "json"
+    }).then((success: IAjaxSuccess): ISessionId => {
+      if (success.xhr.status !== 200) {
+        throw Error('Invalid response');
+      }
+      validateSessionId(success.data);
+      return success.data;
+    });
+  }
+
+  /**
+   * PATCH /api/sessions/[:session_id]
+   *
+   * Rename or move a notebook. If the given name or path are
+   * undefined, then they will not be changed.
+   *
+   */
+  renameNotebook(path: string): Promise<void> {
+    if (path !== undefined) {
+    this._notebookModel.path = path;
+    }
+    return utils.ajaxRequest(this._sessionUrl, {
+      method: "PATCH",
+      dataType: "json",
+      contentType: 'application/json',
+      data: JSON.stringify(this._model)
+    }).then((success: IAjaxSuccess) => {
+      if (success.xhr.status !== 200) {
+        throw Error('Invalid response');
+      }
+      validateSessionId(success.data);
+    });
+  }
+
+  /**
+   * DELETE /api/sessions/[:session_id]
+   *
+   * Kill the kernel and shutdown the session.
+   *
+   */
+  delete(): Promise<void> {
+    if (this._kernel) {
+      this._handleStatus('kernelKilled');
+      this._kernel.close();
+    }
+    return utils.ajaxRequest(this._sessionUrl, {
+      method: "DELETE",
+      dataType: "json"
+    }).then((success: IAjaxSuccess) => {
+      if (success.xhr.status !== 204) {
+        throw Error('Invalid response');
+      }
+      validateSessionId(success.data);
+    });
+  }
+
+  /**
+   * Restart the session by deleting it and the starting it fresh.
+   */
+  restart(options: ISessionOptions): Promise<void> {
+    return this.delete().then(this.start).catch(this.start).then(() => {
+      if (options && options.notebookPath) {
+        this._notebookModel.path = options.notebookPath;
+      }
+      if (options && options.kernelName) {
+        this._kernelModel.name = options.kernelName;
+      }
+      this._kernelModel.id = null;
+    })
+  }
+
+  /**
+   * Get the data model for the session, which includes the notebook path
+   * and kernel (name and id).
+   *
+   */
+  private get _model(): ISessionId {
+    return {
+      id: this._id,
+      notebook: this._notebookModel,
+      kernel: this._kernelModel
+    };
+  }
+
+  /**
+   * Update the data model from the given JSON object, which should
+   * have attributes of `id`, `notebook`, and/or `kernel`. If
+   * provided, the notebook data must include name and path, and the
+   * kernel data must include name and id.
+   *
+   */
+  private _updateModel(data: ISessionId): void {
+    if (data && data.id) {
```

Comment:
data will always be defined and valid as you call `validateSessionID` just before calling that. you can also just call validateSessionId in the first line here, so you are sure that data.id, notebook, and kernel are set IIUC.